lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCacheImpl.java

```diff
@@ -0,0 +1,147 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.TreeMultimap;
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.logsafe.Preconditions;
+
+/**
+ * Notes on concurrency: all public methods in this class are synchronised: this removes any concern that the timestamp
+ * mapping will be modified while also being cleared or read. For processing updates and getting events, this should not
+ * have a performance impact as these methods will be called in a single-threaded manner anyway (via an autobatcher),
+ * but the method to remove entries is not necessarily called as such, and may cause some impact on performance.
+ */
+public final class LockWatchEventCacheImpl implements LockWatchEventCache {
+    private final ClientLockWatchEventLog eventLog;
+    private final HashMap<Long, IdentifiedVersion> timestampMap = new HashMap<>();
+    private final TreeMultimap<IdentifiedVersion, Long> aliveVersions = TreeMultimap.create();
+
+    private LockWatchEventCacheImpl(ClientLockWatchEventLog eventLog) {
+        this.eventLog = eventLog;
+    }
+
+    public static LockWatchEventCacheImpl create() {
+        return create(ClientLockWatchEventLogImpl.create());
+    }
+
+    @VisibleForTesting
+    static LockWatchEventCacheImpl create(ClientLockWatchEventLog eventLog) {
+        return new LockWatchEventCacheImpl(eventLog);
+    }
+
+    @Override
+    public Optional<IdentifiedVersion> lastKnownVersion() {
+        return eventLog.getLatestKnownVersion();
+    }
+
+    @Override
+    public synchronized void processTransactionUpdate(
+            Collection<Long> startTimestamps,
+            LockWatchStateUpdate update) {
+        Optional<IdentifiedVersion> currentVersion = eventLog.getLatestKnownVersion();
+        Optional<IdentifiedVersion> latestVersion = eventLog.processUpdate(update);
+        getEarliestVersion().ifPresent(eventLog::removeOldEntries);
+
+        if (!(latestVersion.isPresent()
+                && currentVersion.isPresent()
+                && latestVersion.get().id().equals(currentVersion.get().id())
+                && update.accept(SuccessVisitor.INSTANCE))) {
+            timestampMap.clear();
+            aliveVersions.clear();
+        }
+
+        latestVersion.ifPresent(
+                version -> startTimestamps.forEach(timestamp -> {
+                    timestampMap.put(timestamp, version);
+                    aliveVersions.put(version, timestamp);
+                }));
+    }
+
+    @Override
+    public synchronized CommitUpdate getCommitUpdate(long startTs, long commitTs, LockToken commitLocksToken) {
```

Comment:
Todo - consider how to handle the `commitLocksToken` - there is a comment above discussing my concern here.